include/ur_api.h

```diff
@@ -847,14 +847,18 @@ urAdapterRetain(
 );
 
 ///////////////////////////////////////////////////////////////////////////////
-/// @brief Get the last adapter specific error.
+/// @brief Get the last adapter specific warning or error.
 ///
 /// @details
 /// To be used after another entry-point has returned
 /// ::UR_RESULT_ERROR_ADAPTER_SPECIFIC in order to retrieve a message describing
 /// the circumstances of the underlying driver error and the error code
 /// returned by the failed driver entry-point.
 ///
+/// * If `pError` is ::UR_RESULT_SUCCESS, ::UR_RESULT_ERROR_ADAPTER_SPECIFIC
```

Comment:
Although I understand it, I wonder if any adapter would add support to return a warning with SUCESS through `urAdapterGetLastError`. Returning a message through `urAdapterGetLastError` has an overhead (strncpy a mesage, keep track of the last message, etc).

this of course could be optional, but I'm concerned this would open the door for UR customers to start requesting warnings, given that the spec allows it, and then having pushback from underlying drivers or adapters that dont want to implement such warning because of performance considerations.

Reply:
For context, this SPEC change is a follow up to discussion that happened in https://github.com/oneapi-src/unified-runtime/pull/989

At the moment, the UR spec says that UR_RESULT_ERROR_ADAPTER_SPECIFIC can be used to return a warning or error but doesn't specify how. The aim to this PR was to clarify what is the correct way to do it. 

There are already adapters using UR_RESULT_ERROR_ADAPTER_SPECIFIC in this way. For example CUDA is doing it for a few entrypoints. In addition, this is the behaviour that SYCL-RT expects.

I understand that there are performance / design implications to returning a warning. The intent was for it to be an optional feature. 

I guess that we have to decide whether we want to support warnings on UR or not. If the answer is yes, then that should be documented in the spec (which this PR aims to do). If the answer is no, then we need to update the spec, SYCL-RT and all the adapters that are using UR_RESULT_ERROR_ADAPTER_SPECIFIC in this way.

Reply:
As an interface layer between higher level languages and lower level drivers I don't think UR should stand in the way of either users requesting more information about why a  feature they want to use isn't working or performance is suboptimal.

As @fabiomestre says, the spec already supports this but does so in an unclear way. I you are concearned about the performance implecations in an adatper @jandres742 then perhaps we should think about ways of opting in to their use rather than being on by default, perahps with an environment variable.

Reply:
thanks @kbenzie , @fabiomestre . I wouldn't even want this to be with env var, as that implies having branches on several places, even on critical paths.

Dont know if we need to clarify on this PR or in the spec that the warnings are optional, but if that's the intention, then I'm ok with this.

Reply:
I think its reasonable to have spec language stating that emitting warnings is optional at the discretion of the adapter maintainers. @fabiomestre if you could update the spec language to clarify that?

Reply:
Added a bullet point to clarify that the warnings are optional in the latest commit.